## 3. Mapping 
Mapping trimmed reads to the reference genome.
To look into script: cat /RAID/Data/linda/all_data/mapping.sh 
Do this before, runnning the command: 

bwa index /home/shangao/Data/EDTA/Ppr/Ppr_instagrall/Ppr_instagrall.polished.fa.mod.MAKER.masked
conda activate uni_tools
mkdir tmp

    for i in 153621_S1 \
    153622_S2 \
    153623_S3 \
    153624_S4 \
    153625_S5 \
    153626_S6
    do
    bwa mem -t 40 -R "@RG\tID:${i}\tSM:${i}\tLB:${i}\tPL:Illumina" /home/shangao/Data/EDTA/Ppr/Ppr_instagrall/Ppr_instagrall.polished.fa.mod.MAKER.masked ${i}.R1_val_1.fq.gz ${i}.R3_val_2.fq.gz > mapped_data/${i}.sam
                #Run Aligntment with masked genome
    samtools view -bS mapped_data//${i}.sam > mapped_data//${i}.bam
                #compress .sam into .bam
    samtools sort mapped_data/${i}.bam -o mapped_data/${i}.sort.bam
                #sort mapped reads according to position in the genome
    samtools index mapped_data/${i}.sort.bam
                #index sorted bams, creates .sort.bam.bai-file
    umi_tools dedup -I mapped_data/${i}.sort.bam --output-stats=deduplicated --paired -S mapped_data/${i}.sort.de.bam --temp-dir=tmp
                #remove all duplications 
    samtools index mapped_data/${i}.sort.de.bam
                #index deduplicated bam
    done

