## 4.1 Creating VCF-files

Creating a GVCF and VCF-File with GATK HaplotypeCaller and GenotypeGVFCs.

    ###software
    gatk=/NVME/Software/popgen/gatk-4.1.9.0/gatk
    ###data
    ref=/home/shangao/Data/EDTA/Ppr/Ppr_instagrall/Ppr_instagrall.polished.fa.mod.MAKER.masked
    bam1=$1
    bam2=$2
    gvcf1=$3
    gvcf2=$4
    mergedgvcf=$5
    vcf=$6
    
    ###build dict for genome
    /NVME/Software/popgen/gatk-4.1.9.0/gatk CreateSequenceDictionary -R /home/shangao/Data/EDTA/Ppr/Ppr_instagrall/Ppr_instagrall.polished.fa.mod.MAKER.masked -O Ppr.FINAL.dict
    
    ###call gvcf
    $gatk HaplotypeCaller \
            -R $ref \
            --emit-ref-confidence GVCF \
            -I $bam1 \
            -O $gvcf1
    
    ###call gvcf
    $gatk HaplotypeCaller \
            -R $ref \
            --emit-ref-confidence GVCF \
            -I $bam2 \
            -O $gvcf2
    
    ###merge them 
    $gatk CombineGVCFs \
        -R $ref \
        -V $gvcf1 \
        -V $gvcf2 \
        -O merged_gvcf/$mergedgvcf

    ###detect SNPs
    $gatk GenotypeGVCFs \
            -R $ref \
            -V $mergedgvcf \
            -O $vcf

    ###compress
    bgzip -f $vcf
    tabix -p vcf ${vcf}.gz
    
    # Filter out only SNPs from VCF 
    $gatk SelectVariants \
    -select-type SNP \
    -V ${vcf}.gz \
    -O ${vcf}.snp.gz
    
    # filter SNPs by parameters
    $gatk VariantFiltration \
    -V ${vcf}.snp.gz \
    --filter-expression "QD <2.0 || MQ <40.0 || FS >60.0 || SOR >3.0 || ReadPosRankSum < -8.0" \
    --filter-name "PASS" \
    -O ${vcf}.snp.f.gz
    
Executed the script with this: 

    sh VCFcalling.sh masked_mapping_data/153621_S1.sort.de.bam masked_mapping_data/153622_S2.sort.de.bam masked_vcf/gvcf/s1.g.vcf masked_vcf/gvcf/s2.g.vcf masked_vcf/merged_vcf/s12.g.vcf masked_vcf/merged_vcf/s12.vcf 
    sh VCFcalling.sh masked_mapping_data/153623_S3.sort.de.bam masked_mapping_data/153624_S4.sort.de.bam masked_vcf/gvcf/s3.g.vcf masked_vcf/gvcf/s4.g.vcf masked_vcf/merged_vcf/s34.g.vcf masked_vcf/merged_vcf/s34.vcf
    sh VCFcalling.sh masked_mapping_data/153625_S5.sort.de.bam masked_mapping_data/15366_S6.sort.de.bam masked_vcf/gvcf/s5.g.vcf masked_vcf/gvcf/s6.g.vcf masked_vcf/merged_vcf/s56.g.vcf masked_vcf/merged_vcf/s56.vcf
