## 4.1 Creating VCF-files

Creating a GVCF and VCF-File with GATK HaplotypeCaller and GenotypeGVFCs.

Script in cat /RAID/Data/linda/all_data/sim_SNP_calling_w_GATK.sh

    ###software
    gatk=/NVME/Software/popgen/gatk-4.1.9.0/gatk
    ###data
    ref=/RAID/Data/mites/genomes/Ppr/version03/Ppr_instagrall.polished.FINAL.fa
    bam=$1
    out1=$2
    out2=$3
    ###build dict for genome
    #/NVME/Software/popgen/gatk-4.1.9.0/gatk CreateSequenceDictionary -R Ppr.FINAL.fa -O Ppr.FINAL.dict
    ###call gvcf
    $gatk HaplotypeCaller \
            -R $ref \
            --emit-ref-confidence GVCF \
            -I $bam \
            -O $out1

    ###detect SNPs
    $gatk GenotypeGVCFs \
            -R $ref \
            -V $out1 \
            -O $out2

    ###compress
    bgzip -f $out2
    tabix -p vcf $out2.gz
    
Executed with: cat /RAID/Data/linda/all_data/mapping.sh

        for i in 153621_S1 \
        153622_S2 \
        153623_S3 \
        153624_S4 \
        153625_S5 \
        153626_S6
        do
        echo "sh sim_SNP_calling_w_GATK.sh mapped_data/${i}.sort.de.bam vcf/${i}.g.vcf vcf/${i}.vcf " > shell/$i.sh
        nohup sh shell/$i.sh > shell/$i.log &
        done

## 4.2 Merging Mother and Daughter & SNP Calling

Combining Mother-daughter pairs with CombineGVCFs. Converting GVCF to VCF with GenotypeGVCFs. Filtering out SNPs with SelectVariant and filtering with VariantFiltration. And then omitting multiallelic SNPs. The following code is an example for the first mother and daughter pair.  

Script in: /RAID/Data/linda/all_data/vcf/s12.filter.sh

    gatk=/NVME/Software/popgen/gatk-4.1.9.0/gatk
    ref=/RAID/Data/mites/genomes/Ppr/version03/Ppr_instagrall.polished.FINAL.fa

    $gatk CombineGVCFs \
        -R $ref \
        -V 153621_S1.g.vcf \
        -V 153622_S2.g.vcf \
        -O merged_gvcf/s12.g.vcf

    $gatk GenotypeGVCFs \
            -R $ref \
            -V merged_gvcf/s12.g.vcf \
            -O merged_gvcf/s12.vcf.gz

    # Filter out only SNPs from VCF 
    $gatk SelectVariants \
    -select-type SNP \
    -V merged_gvcf/s12.vcf.gz \
    -O merged_gvcf/s12.snp.vcf.gz

    # filter SNPs by parameters
    $gatk VariantFiltration \
    -V merged_gvcf/s12.snp.vcf.gz \
    --filter-expression "QD <2.0 || MQ <40.0 || FS >60.0 || SOR >5.0 || ReadPosRankSum < -8.0" \
    --filter-name "PASS" \
    -O merged_gvcf/s12.snp.f.vcf.gz

    # filter for biallelic snps
    $gatk SelectVariants \
        -V merged_gvcf/s12.snp.f.vcf.gz \
        --restrict-alleles-to BIALLELIC \
        -O merged_gvcf/s12.all.vcf.gz
